File: kotlinsensors/src/main/java/com/lexicalninja/kotlinsensors/serializers/CyclingSpeedCadenceSerializer.java
package com.lexicalninja.kotlinsensors.serializers;

import androidx.annotation.Nullable;

/**
 * Created by Saxton on 7/14/16.
 */
public class CyclingSpeedCadenceSerializer {
    class FlagStruct {
        int rawFlags;

        public boolean contains(int flagPosition) {
            return ((rawFlags & (1L << flagPosition)) != 0);
        }

        public FlagStruct() {
        }

        public FlagStruct(int rawFlags) {
            this.rawFlags = rawFlags;
        }
    }

    public class MeasurementFlags {

        final static byte wheelRevolutionDataPresent = 0x1;
        final static byte crankRevolutionDataPresent = 0x2;
    }

    public class Features extends FlagStruct {

        private int rawFeatures;

        public final static int wheelRevolutionDataSupported = 0x1;
        public final static int crankRevolutionDataSupported = 0x2;
        public final static int multipleSensorLocationsSupported = 0x4;

        public Features(int rawFeatures) {
            this.rawFeatures = rawFeatures;
        }

        public boolean isWheelRevolutionDataSupported() {
            return ((rawFeatures & (wheelRevolutionDataSupported & 0xFF)) == wheelRevolutionDataSupported);
        }

        public boolean isCrankRevolutionDataSupported() {
            return ((rawFeatures & (crankRevolutionDataSupported & 0xFF)) == crankRevolutionDataSupported);
        }

        public boolean isMultipleSensorLocationsSupported() {
            return ((rawFeatures & multipleSensorLocationsSupported) == multipleSensorLocationsSupported);
        }
    }

    public class MeasurementData implements CyclingMeasurementData {
        public double timestamp = 0;
        @Nullable
        public Integer cumulativeWheelRevolutions;
        @Nullable
        public Short lastWheelEventTime;
        @Nullable
        public Integer cumulativeCrankRevolutions;
        @Nullable
        public Integer lastCrankEventTime;

        public MeasurementData() {
            super();
        }

        @Override
        public double getTimeStamp() {
            return this.timestamp;
        }

        @Override
        @Nullable
        public Integer getCumulativeCrankRevolutions() {
            return this.cumulativeCrankRevolutions;
        }

        @Override
        @Nullable
        public Integer getCumulativeWheelRevolutions() {
            return cumulativeWheelRevolutions;
        }

        @Override
        @Nullable
        public Integer getLastCrankEventTime() {
            return lastCrankEventTime;
        }

        @Override
        @Nullable
        public Short getLastWheelEventTime() {
            return lastWheelEventTime;
        }
    }

    public static Features readFeatures(byte[] bytes) {
        short rawFeatures = (short) ((bytes[0] & 0xFF) | (bytes[1] | 0xFF) << 8);
        return new CyclingSpeedCadenceSerializer().new Features(rawFeatures);
    }

    public static MeasurementData readMeasurement(byte[] bytes) {
        MeasurementData measurement = new CyclingSpeedCadenceSerializer().new MeasurementData();

        int index = 0;
        byte rawFlags = bytes[index++];

        if ((rawFlags & MeasurementFlags.wheelRevolutionDataPresent) == MeasurementFlags.wheelRevolutionDataPresent) {
            measurement.cumulativeWheelRevolutions = ((int) bytes[index++] & 0xFF) | (((int) bytes[index++] & 0xFF) << 8) | (((int) bytes[index++] & 0xFF) << 16) | (((int) bytes[index++] & 0xFF) << 24);
            measurement.lastWheelEventTime = (short) ((bytes[index++] & 0xff) | ((bytes[index++] & 0xff) << 8));
        }

        if ((rawFlags & MeasurementFlags.crankRevolutionDataPresent) == MeasurementFlags.crankRevolutionDataPresent) {
            measurement.cumulativeCrankRevolutions = ((bytes[index++] & 0xFF) | ((short) bytes[index++] & 0xFF) << 8);
            measurement.lastCrankEventTime = ((bytes[index++] & 0xFF) | ((short) bytes[index] & 0xFF) << 8);

        }

        measurement.timestamp = System.currentTimeMillis();

        return measurement;
    }
}

File: kotlinsensors/src/main/java/com/lexicalninja/kotlinsensors/serializers/CyclingSpeedCadenceSerializer.kt
package com.lexicalninja.kotlinsensors.serializers

import com.lexicalninja.kotlinsensors.FlagStruct

/**
 * Created by Saxton on 7/14/16.
 */
class CyclingSpeedCadenceSerializer {

    object MeasurementFlags {
        const val wheelRevolutionDataPresent: Int = 0
        const val crankRevolutionDataPresent: Int = 1
    }

    class Features(rawFeatures: Int) : FlagStruct(rawFeatures) {
        val isWheelRevolutionDataSupported: Boolean
            get() = wheelRevolutionDataSupported in this

        val isCrankRevolutionDataSupported: Boolean
            get() = crankRevolutionDataSupported in this

        val isMultipleSensorLocationsSupported: Boolean
            get() = multipleSensorLocationsSupported in this

        companion object {
            const val wheelRevolutionDataSupported = 0
            const val crankRevolutionDataSupported = 1
            const val multipleSensorLocationsSupported = 2
        }

    }

    class MeasurementData : CyclingMeasurementData {
        override var cumulativeWheelRevolutions: Int? = null
        override var lastWheelEventTime: Short? = null
        override var cumulativeCrankRevolutions: Int? = null
        override var lastCrankEventTime: Int? = null
        override var timestamp: Double = 0.0

    }

    companion object {
        fun readFeatures(bytes: ByteArray): Features {
            val rawFeatures = (bytes[0].toInt() and 0xFF) or ((bytes[1].toInt() or 0xFF) shl 8)
            return Features(rawFeatures)
        }

        fun readMeasurement(bytes: ByteArray): MeasurementData {
            val measurement = MeasurementData()
            var index = 0
            val rawFlags = FlagStruct(bytes[index++].toInt())
            if (rawFlags.contains(MeasurementFlags.wheelRevolutionDataPresent)) {
                measurement.cumulativeWheelRevolutions =
                    (bytes[index++].toInt() and 0xFF) or
                            ((bytes[index++].toInt() and 0xFF) shl 8) or
                            ((bytes[index++].toInt() and 0xFF) shl 16) or
                            ((bytes[index++].toInt() and 0xFF) shl 24)
                measurement.lastWheelEventTime = ((bytes[index++].toInt() and 0xff) or
                        ((bytes[index++].toInt() and 0xff) shl 8)).toShort()
            }
            if (rawFlags.contains(MeasurementFlags.crankRevolutionDataPresent)) {
                measurement.cumulativeCrankRevolutions = (bytes[index++].toInt() and 0xFF) or
                        ((bytes[index++].toInt() and 0xFF) shl 8)
                measurement.lastCrankEventTime = (bytes[index++].toInt() and 0xFF) or
                        ((bytes[index].toInt() and 0xFF) shl 8)
            }
            measurement.timestamp = System.currentTimeMillis().toDouble()
            return measurement
        }
    }
}